File: src/routes/stream.py
from __future__ import annotations

from pathlib import Path

from fastapi import APIRouter
from fastapi.responses import FileResponse, JSONResponse, RedirectResponse

from src.logger import logger
from src.services.validate import validate_track_id
from src.services.ytdlp import get_desktop_stream_url, get_stream_url, download_audio

router = APIRouter()


@router.get("/api/stream/{track_id}")
async def stream(track_id: str, proxy: bool = False, download: bool = False):
    err = validate_track_id(track_id)
    if err:
        return JSONResponse(status_code=400, content={"error": err})

    logger.info("Stream request: id=%s proxy=%s download=%s", track_id, proxy, download)

    if download:
        file_path = await download_audio(track_id)
        return FileResponse(
            path=file_path,
            media_type="audio/mpeg",
            filename=f"{track_id}.mp3",
        )

    url = await get_stream_url(track_id)
    return RedirectResponse(url=url)


@router.get("/api/stream/{track_id}/desktop")
async def stream_desktop(track_id: str):
    err = validate_track_id(track_id)
    if err:
        return JSONResponse(status_code=400, content={"error": err})

    logger.info("Desktop stream request: id=%s", track_id)
    url = await get_desktop_stream_url(track_id)
    return RedirectResponse(url=url)

File: src/routes/stream.ts
import type { FastifyInstance } from "fastify";
import { stat, createReadStream } from "node:fs";
import { promisify } from "node:util";
import { getStreamUrl, getInfo, downloadAudio } from "../services/ytdlp.js";
import { validateTrackId } from "../services/validate.js";
import { logger } from "../logger.js";

const statAsync = promisify(stat);

export async function streamRoutes(app: FastifyInstance) {
  app.get<{ Params: { id: string }; Querystring: { proxy?: string; download?: string } }>(
    "/api/stream/:id",
    async (req, reply) => {
      const { id } = req.params;

      const validation = validateTrackId(id);
      if (!validation.valid) {
        return reply.status(400).send({ error: validation.reason });
      }

      const useProxy = req.query.proxy === "true";
      const doDownload = req.query.download === "true";

      logger.info({ id, proxy: useProxy, download: doDownload }, "Stream request");

      if (doDownload) {
        const filePath = await downloadAudio(id);
        const st = await statAsync(filePath);
        reply.header("Content-Type", "audio/mpeg");
        reply.header("Content-Length", String(st.size));
        reply.header("Content-Disposition", `attachment; filename="${id}.mp3"`);
        return reply.send(createReadStream(filePath));
      }

      const url = await getStreamUrl(id);
      return reply.redirect(url);
    },
  );
}
